File: server/src/app/user/schemas.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr

from .model import UserRole


class UserCreate(BaseModel):
    """Payload for creating a new user."""
    name: str
    email: EmailStr
    profile_image_url: str | None = None
    supabase_uuid: UUID | None = None
    auth_id: str | None = None
    access_token: str | None = None

class UserUpdate(BaseModel):
    """Payload for partially updating an existing user."""
    name: str | None = None
    email: EmailStr | None = None
    profile_image_url: str | None = None
    auth_id: str | None = None

class UserRead(BaseModel):
    """Read model exposed by user APIs."""
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None
    uuid: str
    name: str
    email: EmailStr
    profile_image_url: str | None = None
    role: UserRole
    auth_id: str | None = None
    
    model_config = ConfigDict(from_attributes=True)

class TeamMembershipRead(BaseModel):
    uuid: str
    user_uuid: str
    team_uuid: str
    role: UserRole
    joined_at: datetime
    class Config:
        orm_mode = True

class UserWithTeamsRead(UserRead):
    team_memberships: list[TeamMembershipRead] | None = None

class TokenRefresh(BaseModel):
    """Payload for requesting a token refresh."""
    refresh_token: str

File: apps/web/src/entities/user/api/schemas.ts
import { z } from 'zod'

export const UserRole = z.enum(['admin', 'user'])
export type UserRole = z.infer<typeof UserRole>

export const TeamMemberRole = z.enum(['owner', 'admin', 'member'])
export type TeamMemberRole = z.infer<typeof TeamMemberRole>

export const UserReadSchema = z.object({
  createdAt: z.string(),
  updatedAt: z.string(),
  deletedAt: z.string().nullable(),
  uuid: z.string(),
  name: z.string(),
  email: z.string().email(),
  profileImageUrl: z.string().nullable(),
  role: UserRole,
  authId: z.string().nullable(),
})
export type UserRead = z.infer<typeof UserReadSchema>

export const TeamMembershipReadSchema = z.object({
  uuid: z.string(),
  userUuid: z.string(),
  teamUuid: z.string(),
  role: TeamMemberRole,
  joinedAt: z.string(),
})
export type TeamMembershipRead = z.infer<typeof TeamMembershipReadSchema>

export const UserWithTeamsReadSchema = UserReadSchema.extend({
  teamMemberships: z.array(TeamMembershipReadSchema).nullable(),
})
export type UserWithTeamsRead = z.infer<typeof UserWithTeamsReadSchema>

export const UserCreateSchema = z.object({
  name: z.string().min(1),
  email: z.string().email(),
  profileImageUrl: z.string().nullable().optional(),
  supabaseUuid: z.string().nullable().optional(),
  authId: z.string().nullable().optional(),
  accessToken: z.string().nullable().optional(),
})
export type UserCreate = z.infer<typeof UserCreateSchema>

export const UserUpdateSchema = z.object({
  name: z.string().min(1).optional(),
  email: z.string().email().optional(),
  profileImageUrl: z.string().nullable().optional(),
  authId: z.string().nullable().optional(),
})
export type UserUpdate = z.infer<typeof UserUpdateSchema>

export const TokenResponseSchema = z.object({
  access_token: z.string(),
  token_type: z.string(),
  expires_in: z.number(),
  refresh_token: z.string().optional(),
})
export type TokenResponse = z.infer<typeof TokenResponseSchema>
